Add GPI function runfile

hedonhermdev opened the pull request:
Fixes #1523 
-Add a function runfile which resembles the function of execfile() in python2.

hedonhermdev committed:
Add GPI function runfile()

Add GPI function runfile()

hedonhermdev commented:
I added the function in GangaCore.Runtime.GPIFunctions. Please let me know if there's a better place for the function to be added.

egede commented on ganga/GangaCore/Runtime/GPIFunctions.py:
```diff
@@ -93,3 +95,9 @@ def convert_merger_to_postprocessor(j):
         stripProxy(j).postprocessors = mp
 
+def runfile(path_to_file):
+    """ A wrapper for the runpy.run_path() function. """
```
I suggest adding an example of usage here as well. Maybe just the simple usage `runpy('myfile.py')` but then also add an example of what can be in `myfile.py`. The important thing to illustrate is how it behaves with the namespace of the `Ganga` session. So can I have 

```
j = Job()
j.submit()
```
in the file? And comment on if the object `j` in this case will be visible at the main prompt after running.

egede commented on ganga/GangaCore/Runtime/GPIFunctions.py:
```diff
@@ -93,3 +95,9 @@ def convert_merger_to_postprocessor(j):
 
+def runfile(path_to_file):
+    """ A wrapper for the runpy.run_path() function. """
+    if not type(path_to_file) == str:
```
Avoid direct type comparisons. Always use `if not isinstance(path_to_file, str):` instead as somebody might in principle have a class that inherits from `str`.

hedonhermdev committed:
Modify docstring and change typechecking in runfile()

egede commented on ganga/GangaCore/Runtime/GPIFunctions.py:
```diff
@@ -93,3 +95,21 @@ def convert_merger_to_postprocessor(j):
+
+    then you have to run the function with init_globals=globals() in the Ganga interpreter like this
+
+     Ganga In[]: runfile('myfile.py', init_globals=globals())
```
As we are exporting this to the Ganga API, I would make the `init_globals=globals()` the default. I actually think it should not even be an option. If a user wants anything else, it is not really a part of the Ganga API, and they can just import runpy themselves.

hedonhermdev replied:
Makes sense, but I had trouble making it the default as setting init_globals=globals() in the function sets it to the globals of that file rather than the namespace of the GPI interpreter. Any ideas?

egede replied:
I am not quite sure what will work here. Can you try something like

```
    from GangaCore.GPI import *
    return runpy.run_path(path_to_file, init_globals=globals())
```
where the idea is that we populate the local namespace with what has been exposed to the API of Ganga (called the GPI).

egede replied:
Did you try this?

egede reviewed (commented):
Please add a test case that demonstrates that this is working. It should be placed in a new file in `GangaCore/test/GPI`. You can run the test yourself with `pytest mytest.py`. It will also run as part of the unit testing for every push made to this pull request.

hedonhermdev committed:
Add test for runfile

hedonhermdev commented:
How can I spawn a Ganga Interpreter in a GangaUnitTest?

egede commented:
> How can I spawn a Ganga Interpreter in a GangaUnitTest?

See https://github.com/ganga-devs/ganga/blob/develop/ganga/GangaCore/test/GPI/TestArgSplitter.py but I do not think that you actually want this?

egede committed:
Merge branch 'develop' into run

mesmith75 committed:
Merge branch 'develop' into run

hedonhermdev committed:
Fix runfile and delete test.

hedonhermdev committed:
Add test for runfile

egede commented on ganga/GangaCore/Runtime/GPIFunctions.py:
```diff
@@ -93,3 +96,22 @@ def convert_merger_to_postprocessor(j):
+
+    then you have to run the function with init_globals=globals() in the Ganga interpreter like this
+
+     Ganga In[]: runfile('myfile.py', init_globals=globals());
```
The function no longer takes the `init_globals` argument, so this line should be fixed.

egede commented on ganga/GangaCore/test/GPI/TestRunfile.py:
```diff
@@ -0,0 +1,11 @@
+            f.write('j = Job()\nj.submit()')
+            f.close()
+        runfile('test_file.py')
+        os.remove('test_file.py')
```
Add newline at end of file

egede reviewed (changes requested):
Tests is fine. See other small comments. Nearly ready.

hedonhermdev committed:
Minor fixes.

hedonhermdev committed:
Update docstring

egede committed:
Remove return value

Took away the return statement from the `run_path` command to avoid that user get dictionary back. Updated documentation to reflect this.

hedonhermdev commented:
Why is the CI failing? I cannot see the details for the Jenkins CI.

egede commented:
> Why is the CI failing? I cannot see the details for the Jenkins CI.

Not quite sure. It runs again now as I made a few more tiny changes. We have had some intermittent problems lately with the CI failing during the build stage of the docker container (that is used for the testing).

egede commented:
> > Why is the CI failing? I cannot see the details for the Jenkins CI.
> 
> Not quite sure. It runs again now as I made a few more tiny changes. We have had some intermittent problems lately with the CI failing during the build stage of the docker container (that is used for the testing).

And it worked now.

egede commented:
Please delete this branch and start on a new one if you make other contributions. Many thanks for the work.
